"runtime error: invalid memory address or nil pointer dereference" reported during tptest

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

tptest 

### 2. What did you expect to see? (Required)
no unexpected error

### 3. What did you see instead (Required)
[2023/03/27 11:37:05.505 +00:00] [ERROR] [executor.go:1910] ["resultPuller panicked"] [conn=1844186664012886753] [recover="runtime error: invalid memory address or nil pointer dereference"] [stack="github.com/pingcap/tidb/executor.(*UnionExec).resultPuller.func1\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:1910\nruntime.gopanic\n\t/usr/local/go/src/runtime/panic.go:884\nruntime.panicmem\n\t/usr/local/go/src/runtime/panic.go:260\nruntime.sigpanic\n\t/usr/local/go/src/runtime/signal_unix.go:837\ngithub.com/tikv/client-go/v2/txnkv/transaction.(*KVTxn).DoneAggressiveLocking\n\t/go/pkg/mod/github.com/tikv/client-go/v2@v2.0.7-0.20230317032622-884a634378d4/txnkv/transaction/txn.go:745\ngithub.com/pingcap/tidb/store/driver/txn.(*tikvTxn).exitFairLockingIfInapplicable\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/driver/txn/txn_driver.go:368\ngithub.com/pingcap/tidb/store/driver/txn.(*tikvTxn).LockKeysFunc\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/store/driver/txn/txn_driver.go:92\ngithub.com/pingcap/tidb/session.(*LazyTxn).LockKeysFunc\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/txn.go:473\ngithub.com/pingcap/tidb/session.(*LazyTxn).LockKeys\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/session/txn.go:448\ngithub.com/pingcap/tidb/executor.doLockKeys\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:1341\ngithub.com/pingcap/tidb/executor.(*SelectLockExec).Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:1273\ngithub.com/pingcap/tidb/executor.Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:326\ngithub.com/pingcap/tidb/executor.(*ProjectionExec).unParallelExecute\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/projection.go:196\ngithub.com/pingcap/tidb/executor.(*ProjectionExec).Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/projection.go:183\ngithub.com/pingcap/tidb/executor.Next\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:326\ngithub.com/pingcap/tidb/executor.(*UnionExec).resultPuller\n\t/home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/tidb/executor/executor.go:1940"]
[2023/03/27 11:37:05.507 +00:00] [WARN] [session.go:2202] ["run statement failed"] [conn=1844186664012886753] [schemaVersion=34580] [error="runtime error: invalid memory address or nil pointer dereference"] [session="{\n  \"currDBName\": \"multitable_dsn1_00\",\n  \"id\": 1844186664012886753,\n  \"status\": 3,\n  \"strictMode\": true,\n  \"txn\": \"440380168717729819\",\n  \"user\": {\n    \"Username\": \"root\",\n    \"Hostname\": \"10.233.83.78\",\n    \"CurrentUser\": false,\n    \"AuthUsername\": \"root\",\n    \"AuthHostname\": \"%\",\n    \"AuthPlugin\": \"mysql_native_password\"\n  }\n}"]

### 4. What is your TiDB version? (Required)

1e5b929e16fe5892cdfab18a1ef0268790766060

--- a/DEPS.bzl
+++ b/DEPS.bzl
@@ -4101,9 +4101,8 @@ def go_deps():
         name = "com_github_tikv_client_go_v2",
         build_file_proto_mode = "disable_global",
         importpath = "github.com/tikv/client-go/v2",
-        replace = "github.com/MyonKeminta/client-go/v2",
-        sum = "h1:K+yZiivn5i/ZhvfRCxATR8vXVJEV+xXYn+ZmJcW1+tg=",
-        version = "v2.0.0-20230328062815-eb5212c07f03",
+        sum = "h1:oVFdu5s0ASRiKYqo2OmYb1e4OhQ6xYpcqqrwwpp1Kss=",
+        version = "v2.0.7-0.20230328075239-e72043db9e79",
     )
     go_repository(
         name = "com_github_tikv_pd",

--- a/DEPS.bzl
+++ b/DEPS.bzl
@@ -4101,8 +4101,9 @@ def go_deps():
         name = "com_github_tikv_client_go_v2",
         build_file_proto_mode = "disable_global",
         importpath = "github.com/tikv/client-go/v2",
-        sum = "h1:m5Y7tBW5Rq8L1ANxibitBa/DInDy3hA2Qvk1Ys9u1NU=",
-        version = "v2.0.7-0.20230317032622-884a634378d4",
+        replace = "github.com/MyonKeminta/client-go/v2",
+        sum = "h1:K+yZiivn5i/ZhvfRCxATR8vXVJEV+xXYn+ZmJcW1+tg=",
+        version = "v2.0.0-20230328062815-eb5212c07f03",
     )
     go_repository(
         name = "com_github_tikv_pd",

--- a/go.mod
+++ b/go.mod
@@ -280,5 +280,6 @@ replace (
 	// fix potential security issue(CVE-2020-26160) introduced by indirect dependency.
 	github.com/dgrijalva/jwt-go => github.com/form3tech-oss/jwt-go v3.2.6-0.20210809144907-32ab6a8243d7+incompatible
 	github.com/pingcap/tidb/parser => ./parser
+	github.com/tikv/client-go/v2 => github.com/MyonKeminta/client-go/v2 v2.0.0-20230328062815-eb5212c07f03
 	go.opencensus.io => go.opencensus.io v0.23.1-0.20220331163232-052120675fac
 )

--- a/go.mod
+++ b/go.mod
@@ -92,7 +92,7 @@ require (
 	github.com/stretchr/testify v1.8.2
 	github.com/tdakkota/asciicheck v0.1.1
 	github.com/tiancaiamao/appdash v0.0.0-20181126055449-889f96f722a2
-	github.com/tikv/client-go/v2 v2.0.7-0.20230317032622-884a634378d4
+	github.com/tikv/client-go/v2 v2.0.7-0.20230328075239-e72043db9e79
 	github.com/tikv/pd/client v0.0.0-20230324043643-8c21dfc56c74
 	github.com/timakin/bodyclose v0.0.0-20221125081123-e39cf3fc478e
 	github.com/twmb/murmur3 v1.1.6
@@ -280,6 +280,5 @@ replace (
 	// fix potential security issue(CVE-2020-26160) introduced by indirect dependency.
 	github.com/dgrijalva/jwt-go => github.com/form3tech-oss/jwt-go v3.2.6-0.20210809144907-32ab6a8243d7+incompatible
 	github.com/pingcap/tidb/parser => ./parser
-	github.com/tikv/client-go/v2 => github.com/MyonKeminta/client-go/v2 v2.0.0-20230328062815-eb5212c07f03
 	go.opencensus.io => go.opencensus.io v0.23.1-0.20220331163232-052120675fac
 )

--- a/store/driver/txn/txn_driver.go
+++ b/store/driver/txn/txn_driver.go
@@ -79,7 +79,6 @@ func (txn *tikvTxn) CacheTableInfo(id int64, info *model.TableInfo) {
 
 func (txn *tikvTxn) LockKeys(ctx context.Context, lockCtx *kv.LockCtx, keysInput ...kv.Key) error {
 	keys := toTiKVKeys(keysInput)
-	txn.exitFairLockingIfInapplicable(ctx, keys)
 	err := txn.KVTxn.LockKeys(ctx, lockCtx, keys...)
 	if err != nil {
 		return txn.extractKeyErr(err)
@@ -89,7 +88,6 @@ func (txn *tikvTxn) LockKeys(ctx context.Context, lockCtx *kv.LockCtx, keysInput
 
 func (txn *tikvTxn) LockKeysFunc(ctx context.Context, lockCtx *kv.LockCtx, fn func(), keysInput ...kv.Key) error {
 	keys := toTiKVKeys(keysInput)
-	txn.exitFairLockingIfInapplicable(ctx, keys)
 	err := txn.KVTxn.LockKeysFunc(ctx, lockCtx, fn, keys...)
 	if err != nil {
 		return txn.extractKeyErr(err)
@@ -355,18 +353,6 @@ func (txn *tikvTxn) SetAssertion(key []byte, assertion ...kv.FlagsOp) error {
 
 func (txn *tikvTxn) UpdateMemBufferFlags(key []byte, flags ...kv.FlagsOp) {
 	txn.GetUnionStore().GetMemBuffer().UpdateFlags(key, getTiKVFlagsOps(flags)...)
-}
-
-func (txn *tikvTxn) exitFairLockingIfInapplicable(ctx context.Context, keys [][]byte) {
-	if len(keys) > 1 && txn.IsInAggressiveLockingMode() {
-		// Only allow fair locking if it only needs to lock one key. Considering that it's possible that a
-		// statement causes multiple calls to `LockKeys` (which means some keys may have been locked in fair
-		// locking mode), here we exit fair locking mode by calling DoneFairLocking instead of cancelling.
-		// Then the previously-locked keys during execution in this statement (if any) will be turned into the state
-		// as if they were locked in normal way.
-		// Note that the issue https://github.com/pingcap/tidb/issues/35682 also exists here.
-		txn.KVTxn.DoneAggressiveLocking(ctx)
-	}
 }
 
 func (txn *tikvTxn) generateWriteConflictForLockedWithConflict(lockCtx *kv.LockCtx) error {

